src/core/layout/qgslayoutitemmap.cpp

```diff
@@ -1155,6 +1155,14 @@ QgsExpressionContext QgsLayoutItemMap::createExpressionContext() const
   scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "map_crs_definition" ), mapCrs.toProj4(), true ) );
   scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "map_units" ), QgsUnitTypes::toString( mapCrs.mapUnits() ), true ) );
 
+  QVariantList layers;
+  QList<QgsMapLayer *> layersInMap = layersToRender( &context );
+  for ( QgsMapLayer *layer : layersInMap )
+  {
+    layers << layer->id();
+  }
+  scope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "map_layers" ), layers, true ) );
```

Comment:
Ok - this is tricky, but you'll need to call layersToRender *after* adding the scope to the context, and only add the map_layers variable right at the end of this function before you return the context.

Otherwise, someone who uses data defined map layers relying on one of the other variables in this scope (such as @map_extent, @map_scale, etc) will get incorrect layers included in the @map_layers variable, because the call layersToRender hasn't had that information available in the context.

I'd also suggest adding a big warning note that the call to layersToRender needs to be right at the end of this function, after adding all other variables to the context.